patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch

```diff
@@ -0,0 +1,426 @@
+-
++#ifndef OPENSSL_IS_BORINGSSL
+   if (!param_ctx ||
+       EVP_PKEY_paramgen_init(param_ctx.get()) <= 0 ||
+       EVP_PKEY_CTX_set_dsa_paramgen_bits(
+@@ -49,7 +49,7 @@ EVPKeyCtxPointer DsaKeyGenTraits::Setup(DsaKeyPairGenConfig* params) {
+       return EVPKeyCtxPointer();
+     }
+   }
+-
++#endif
+   if (EVP_PKEY_paramgen(param_ctx.get(), &raw_params) <= 0)
+     return EVPKeyCtxPointer();
+ 
+diff --git a/src/crypto/crypto_ec.cc b/src/crypto/crypto_ec.cc
+index ea4c70ad5d8c844860ba3480fc7ef4205f0a3cdc..cdf8dd47d6e2a5894066cec01fbe347af079ec22 100644
+--- a/src/crypto/crypto_ec.cc
++++ b/src/crypto/crypto_ec.cc
+@@ -314,7 +314,7 @@ void ECDH::SetPrivateKey(const FunctionCallbackInfo<Value>& args) {
+     return THROW_ERR_CRYPTO_OPERATION_FAILED(env,
+         "Failed to set generated public key");
+ 
+-  EC_KEY_copy(ecdh->key_.get(), new_key.get());
++  ecdh->key_.reset(EC_KEY_dup(new_key.get()));
+   ecdh->group_ = EC_KEY_get0_group(ecdh->key_.get());
+ }
+ 
+diff --git a/src/crypto/crypto_hkdf.cc b/src/crypto/crypto_hkdf.cc
+index 0aa96ada47abe4b66fb616c665101278bbe0afb6..1e9a4863c5faea5f6b275483ca16f3a6e8dac25b 100644
+--- a/src/crypto/crypto_hkdf.cc
++++ b/src/crypto/crypto_hkdf.cc
+@@ -101,6 +101,7 @@ bool HKDFTraits::DeriveBits(
+     Environment* env,
+     const HKDFConfig& params,
+     ByteSource* out) {
++#ifndef OPENSSL_IS_BORINGSSL
+   EVPKeyCtxPointer ctx =
+       EVPKeyCtxPointer(EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, nullptr));
+   if (!ctx ||
+@@ -132,6 +133,9 @@ bool HKDFTraits::DeriveBits(
+ 
+   *out = std::move(buf);
+   return true;
++#else
++  return false;
++#endif
+ }
+ 
+ void HKDFConfig::MemoryInfo(MemoryTracker* tracker) const {
+diff --git a/src/crypto/crypto_random.cc b/src/crypto/crypto_random.cc
+index b24f8f32136ffaed54310d5dc02e57b0f69450d6..50a6663966cdb147a702df21240fa449850c3549 100644
+--- a/src/crypto/crypto_random.cc
++++ b/src/crypto/crypto_random.cc
+@@ -8,6 +8,7 @@
+ #include "v8.h"
+ 
+ #include <openssl/bn.h>
++#include <openssl/rand.h>
+ 
+ namespace node {
+ 
+@@ -149,7 +150,7 @@ Maybe<bool> RandomPrimeTraits::AdditionalConfig(
+ 
+   params->bits = bits;
+   params->safe = safe;
+-  params->prime.reset(BN_secure_new());
++  params->prime.reset(BN_new());
+   if (!params->prime) {
+     THROW_ERR_CRYPTO_OPERATION_FAILED(env, "could not generate prime");
+     return Nothing<bool>();
+diff --git a/src/crypto/crypto_rsa.cc b/src/crypto/crypto_rsa.cc
+index 5fa91cce1a6ad2bc1167e20a4dadcfdfc2343440..6f1bad3f129066eddcbddf32849b6d4832b62ae9 100644
+--- a/src/crypto/crypto_rsa.cc
++++ b/src/crypto/crypto_rsa.cc
+@@ -210,7 +210,10 @@ WebCryptoCipherStatus RSA_Cipher(
+   if (label_len > 0) {
+     void* label = OPENSSL_memdup(params.label.get(), label_len);
+     CHECK_NOT_NULL(label);
+-    if (EVP_PKEY_CTX_set0_rsa_oaep_label(ctx.get(), label, label_len) <= 0) {
++    if (EVP_PKEY_CTX_set0_rsa_oaep_label(
++      ctx.get(),
++      reinterpret_cast<unsigned char*>(label),
```

Comment:
nit: the underlying type in boringssl here is `uint8_t`, so perhaps
```suggestion
+      reinterpret_cast<uint8_t*>(label),
```

Reply:
I upstreamed it [here](https://github.com/nodejs/node/pull/38926) and Node.js uses `static_cast` by convention so let's switch to that since it's what we'll eventually receive from upstream?